posenet/src/posenet_model.ts

```diff
@@ -167,13 +204,97 @@ export class PoseNet {
 
+    const scaleY = height / resizedHeight;
+    const scaleX = width / resizedWidth;
+
     heatmapScores.dispose();
     offsets.dispose();
 
+    return scalePose(pose, scaleY, scaleX);
+  }
+
+  /**
+   * Infer through PoseNet, and estimates a single pose, segmentation mask, and
+   * colored part image using the outputs. This does standard ImageNet
+   * pre-processing before inferring through the model. The image should pixels
+   * should have values [0-255]. This method returns a single pose, a
+   * segmentation mask, and colored part map imagee.
+   *
+   * @param input ImageData|HTMLImageElement|HTMLCanvasElement|HTMLVideoElement)
+   * The input image to feed through the network.
+   *
+   * @param imageScaleFactor A number between 0.2 and 1. Defaults to 0.50. What
+   * to scale the image by before feeding it through the network.  Set this
+   * number lower to scale down the image and increase the speed when feeding
+   * through the network at the cost of accuracy.
+   *
+   * @param flipHorizontal.  Defaults to false.  If the poses should be
+   * flipped/mirrored  horizontally.  This should be set to true for videos
+   * where the video is by default flipped horizontally (i.e. a webcam), and you
+   * want the poses to be returned in the proper orientation.
+   *
+   * @param outputStride the desired stride for the outputs.  Must be 32, 16,
+   * or 8. Defaults to 16. The output width and height will be will be
+   * (inputDimension - 1)/outputStride + 1
+   *
+   * @param segmentationThreshold The minimum that segmentation values must have
+   * to be considered part of the person.  Affects the generation of the
+   * segmentation mask and the clipping of the colored part image.
+   *
+   * @param partColors An array of rgb color values indexed by part channel id
+   * that are used to generate the part map image
+   *
+   * @return A single pose with a confidence score, which contains an array of
+   * keypoints indexed by part id, each with a score and position.  The
+   * positions of the keypoints are in the same scale as the original image
+   */
+  async estimateSinglePoseWithSegmentation(
+      input: InputType, imageScaleFactor = 0.5, flipHorizontal = false,
+      outputStride: OutputStride = 16, segmentationThreshold = 0.5,
+      partColors: Array<[number, number, number]>):
+      Promise<PoseWithSegmentationMaskAndColoredPartImage> {
+    assertValidOutputStride(outputStride);
+    assertValidScaleFactor(imageScaleFactor);
+
+    const [height, width] = getInputTensorDimensions(input);
+    const resizedHeight =
+        getValidResolution(imageScaleFactor, height, outputStride);
+    const resizedWidth =
+        getValidResolution(imageScaleFactor, width, outputStride);
+
+    const {heatmapScores, offsets, segmentScores, partHeatmapScores} =
+        tf.tidy(() => {
+          const inputTensor =
+              toInputTensor(input, resizedHeight, resizedWidth, flipHorizontal);
```

Comment:
The new model works best with input crop of size 353x257 (pixels), as that's what it was trained on, so it is necessary to add a functionality to crop a sub region of the `inputTensor` and resize the sub-region to 353x257 pixels before feeding into the network. To prevent introducing aspect-ratio distortion, the sub-region should have the aspect ratio of (353 : 257) before being resized to 353x257 (pixels).

Reply:
Will do.